### PR TITLE
[mellanox] Backport patches for mellanox hw-mgmt V.7.0020.1300

### PR DESCRIPTION
**Pull request by alexrallen**
Here are the new patches. Some of the other patches are also reformatted and renamed to help out with internal consistency on our end, no functionality has been changed. 

Patch | Upstream|
|------------|-----------------------------|
|0043-platform-mellanox-mlxreg-io-Fix-argument-base-in-kst.patch| https://github.com/torvalds/linux/commit/452dcfab995437888106a5af6730363f67efdf79|
|0044-platform-mellanox-mlxreg-io-Fix-read-access-of-n-byt.patch| https://github.com/torvalds/linux/commit/5fd56f11838da1580cc77743c08cf02c9cd48380|
|0045-i2c-mlxcpld-Fix-criteria-for-frequency-setting.patch| https://github.com/torvalds/linux/commit/52f57396c75acd77ebcdf3d20aed24ed248e9f79|
|0046-i2c-mlxcpld-Reduce-polling-time-for-performance-impr.patch| https://github.com/torvalds/linux/commit/669b2e4aa1a869def4dc207ea084fdd77366d646|
|0047-i2c-mlxcpld-Allow-flexible-polling-time-setting-for-.patch| https://github.com/torvalds/linux/commit/712d6617d0a29dc4d4da361247720cdf156ae7bb|
|0048-hwmon-pmbus-mp2975-Add-missed-POUT-attribute-for-pag.patch| https://github.com/torvalds/linux/commit/2292e2f685cd5c65e3f47bbcf9f469513acc3195|
|0052-hwmon-mlxreg-fan-Return-non-zero-value-when-fan-curr.patch| https://github.com/torvalds/linux/commit/e6fab7af6ba1bc77c78713a83876f60ca7a4a064|
|0053-mlxsw-core-Avoid-creation-virtual-hwmon-objects-by-t.patch| Already in SONiC 202106|
|0054-mlxsw-minimal-Simplify-method-of-modules-number-dete.patch| Already in SONiC 202106|
|0055-platform_data-mlxreg-Add-new-type-to-support-modular.patch| https://github.com/torvalds/linux/commit/aafa1cafedca7c64d6a43cd21488d28d1d1be884|
|0056-platform-x86-mlx-platform-Add-initial-support-for-ne.patch| https://github.com/torvalds/linux/commit/a5d8f57edfb44a3f036152258d844c21c0436382|
|0057-platform-mellanox-mlxreg-hotplug-Extend-logic-for-ho.patch| https://github.com/torvalds/linux/commit/bb1023b6da379985ff888dcd7bc601735d02267a|
|0058-platform-x86-mlx-platform-Configure-notifier-callbac.patch| https://github.com/torvalds/linux/commit/67eb006cc1d167db33de2fe87988198b9806e794|
|0059-platform-mellanox-mlxreg-io-Extend-number-of-hwmon-a.patch| https://github.com/torvalds/linux/commit/bbfd79c681706e2be8cdf95e2711b9ff9c669ae4|
|0060-platform_data-mlxreg-Add-new-field-for-secured-acces.patch| https://github.com/torvalds/linux/commit/9d93d7877c9158d059028681b66a0c192f85c64d|
|0061-platform-mellanox-mlxreg-lc-Add-initial-support-for-.patch| https://github.com/torvalds/linux/commit/62f9529b8d5c87b89c3d9698405014bc8f370fbd|
|0062-Documentation-ABI-Add-new-attributes-for-mlxreg-io-s.patch| https://github.com/torvalds/linux/commit/5b0a315c3db548646e2cc9707d8a82d818a3c956|
|0063-Documentation-ABI-Add-new-line-card-attributes-for-m.patch| https://github.com/torvalds/linux/commit/164e32717cbde87a2df27183b985631b49701fad|
|0064-hwmon-mlxreg-fan-Extend-the-maximum-number-of-tachom.patch| https://github.com/torvalds/linux/commit/bc8de07e8812548cc19161af3a2b83849ff03045|
|0065-platform-x86-mlx-platform-Extend-FAN-and-LED-config-.patch| https://github.com/torvalds/linux/commit/9045512ca6cdb221cd1ed32d483eac3c30c53bed|
|0066-platform-x86-mlx-platform-Add-new-attributes-for-Cof.patch| https://github.com/torvalds/linux/commit/4289fd4ad43afaed6d1515c573efe9e42a83219f|
|0068-mlxsw-core-Initialize-switch-driver-last.patch| https://github.com/torvalds/linux/commit/3d7a6f677905ea74c060709a0c5fe7b3df1eaae3|
|0069-mlxsw-core-Remove-mlxsw_core_is_initialized.patch| https://github.com/torvalds/linux/commit/25a91f835a7b0964455594d11253060bf5b480b4|
|0070-mlxsw-core_env-Defer-handling-of-module-temperature-.patch| https://github.com/torvalds/linux/commit/163f3d2dd01cf9b8f58b56fe94ce517e468385a9|
|0071-mlxsw-core_env-Convert-module_info_lock-to-a-mutex.patch| https://github.com/torvalds/linux/commit/bd6e43f5953d39d5cb06041d5ae3e9e5470336ef|
|0072-mlxsw-Track-per-module-port-status.patch| https://github.com/torvalds/linux/commit/896f399be0780549703d23fabe6e6bf40c661475|
|0073-mlxsw-reg-Add-fields-to-PMAOS-register.patch| https://github.com/torvalds/linux/commit/8f4ebdb0a2744557a95810787126eb0555e7424b|
|0074-mlxsw-Make-PMAOS-pack-function-more-generic.patch| https://github.com/torvalds/linux/commit/8f4ebdb0a2744557a95810787126eb0555e7424b|
|0075-mlxsw-Add-support-for-transceiver-modules-reset.patch| https://github.com/torvalds/linux/commit/49fd3b645de8c2cc124c284b2be8985c50e76ee0|
|0076-ethtool-Add-ability-to-control-transceiver-modules-p.patch| https://github.com/torvalds/linux/commit/353407d917b2d87cd8104a0453d012439c6ca4be|
|0077-mlxsw-reg-Add-Port-Module-Memory-Map-Properties-regi.patch| https://github.com/torvalds/linux/commit/f10ba086f7e30904ea7cafe7ba922bafc65ad9ad|
|0078-mlxsw-reg-Add-Management-Cable-IO-and-Notifications-.patch| https://github.com/torvalds/linux/commit/fc53f5fb8037d3d29a90c3779d961982ce99e380|
|0079-mlxsw-Add-ability-to-control-transceiver-modules-pow.patch| https://github.com/torvalds/linux/commit/0455dc50bccab9286662f847f560cde6a648802d|
|0080-ethtool-Add-transceiver-module-extended-states.patch| https://github.com/torvalds/linux/commit/3dfb51126064b594470b9c0b278188fbc9194709|
|0081-platform-x86-mlx-platform-Add-support-for-multiply-c.patch| https://github.com/torvalds/linux/commit/249606d37d205eb2f9a6f2c8ecb8bd77b53e5d3e|
|0091-platform-x86-mlx-platform-Add-support-for-new-system.patch| https://github.com/torvalds/linux/commit/4616e54795cc8183ecf2a6cef6c7a9091cb1ae56|
|0092-platform-mellanox-mlxreg-lc-fix-error-code-in-mlxreg.patch| https://github.com/torvalds/linux/commit/287273a80be5d45d59d1742557cab69cabf0ef3c|
|0093-hwmon-mlxreg-fan-Extend-driver-to-support-multiply-P.patch| https://github.com/torvalds/linux/commit/150f1e0c6fa886e18e35594ae2ba5c81b5df1898|
|0094-hwmon-mlxreg-fan-Extend-driver-to-support-multiply-c.patch| https://github.com/torvalds/linux/commit/d7efb2ebc7b3c952104b9ebfbf88da97ea99a0a0|
|0095-hwmon-mlxreg-fan-Fix-out-of-bounds-read-on-array-fan.patch| https://github.com/torvalds/linux/commit/000cc5bc49aafc83a131bab2becf9922f5eec658|
|0096-hwmon-mlxreg-fan-Modify-PWM-connectivity-validation.patch| https://github.com/torvalds/linux/commit/b1c24237341f6fb910c6cb15489222a9f47258d6|
|0097-hwmon-mlxreg-fan-Support-distinctive-names-per-diffe.patch| https://github.com/torvalds/linux/commit/b2be2422c0c98b00f21c4331e2d9342bd58bfdba|

### CODE REVIEW

**Review by dgsudharsan — APPROVED** [review]

**Review by paulmenzel — CHANGES_REQUESTED** [review]
Thank you for the pull/merge request. Unfortunately, it’s hard to review on GitHub. Could do a separate clean-up pull/merge request? Also, the some of the commits have wrong and/or misleading commit message summaries and lack description.

1.  Where can more information about *Mellanox hw-mgmt V.7.0020.1300* be found?
2.  *Update patches* changes Kconfig variables. For each Kconfig variable change, I’d prefer one commit with a commit message saying for what model the new option/driver is needed.
3.  *Fix some patch formatting and fix pmbus issue* needs to be split up, and the PMBus issue be elaborated on.
4.  *Remove config options for not upstreamed patched* reverts changes done in *Update patches*? Please squash.
5.  *Fix regmap access for linecards* lacks the problem description, and why the fix works. Maybe also prefix the commit message summary, so it’s clear what devices are affected. Looks like it could be a separate pull/merge request?
6.  *Another formatting change plus fix more pmbus issues* should be more precise, and the PMBus fix should be a separate commit.

**Comment by liat-grozovik** [comment]
hw mgmt can is public and can be found in https://github.com/Mellanox/hw-mgmt
As for the list of patches we are based on you can further look at https://github.com/Mellanox/hw-mgmt/tree/V.7.0020.1300_BR/recipes-kernel/linux/linux-5.10. It should be as closed as possible but we do have some changes to align to SONiC Linux Kernel asks.
The patches are aligned with the official list of patches. we can provide a feedback to the hw mgmt owner but I dont want to change the patches and dont have an alignment if this is possible. If need to extend the description we will do so.

Also, I don't quite understand what is a clean up pull/merge request? what is missing? @alexrallen is it clear to you on the request?

File: patch/0052-hwmon-mlxreg-fan-Return-non-zero-hwmon-mlxreg-f.patch [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+@@ -329,6 +329,12 @@ static int mlxreg_fan_set_cur_state(struct thermal_cooling_device *cdev,
+ 	 * overwritten.
+ 	 */
+ 	if (state >= MLXREG_FAN_SPEED_MIN && state <= MLXREG_FAN_SPEED_MAX) {
++		/*
++		 * This is configuration change, which is only supported through sysfs.
++		 * For configuration non-zero value is to be returned to avoid thermal
++		 * statistics update.
++		 */
++		config = 0; /*1*/;
```
Comment (saiarcot895) [inline patch/0052-hwmon-mlxreg-fan-Return-non-zero-hwmon-mlxreg-f.patch:96]: The patch description talks about returning a non-zero code if it's a config change, but the `config` variable always returns 0. Am I missing something here?

Comment (alexrallen) [reply patch/0052-hwmon-mlxreg-fan-Return-non-zero-hwmon-mlxreg-f.patch:96]: @vadimp-nvidia Can you please take a look at this?

**Comment by saiarcot895** [comment]
One minor nitpick is that in the patches, the subject line is somewhat repeated (the first couple words are there, then it starts over). If it's easy to fix, then that would be appreciated.

This PR as a whole will be squash-merged as a single commit, so I'm fine with the individual commits here fixing things up. However, I did notice in 24b0956 that the write function call was commented out (which makes the changes done to `regval` in the preceding lines effectively a no-op). Is there a reason why it has to be commented out for SONiC?

For the kconfig changes, as Paul mentioned, at least for the sensor/hardware device additions, is it possible to add a comment saying what device they're for?

File: patch/0052-hwmon-mlxreg-fan-Return-non-zero-hwmon-mlxreg-f.patch (hunk 1 above)
Comment (vadimp-nvidia) [reply patch/0052-hwmon-mlxreg-fan-Return-non-zero-hwmon-mlxreg-f.patch:96]: This part of driver supports special functionality to allow set minimum limit for fan speed. There is upstream work in progress to modify this feature. This work affects three different fan drivers, used by Nvidia systems. Until all these drivers are not aligned it is necessary to comment out this line to satisfy user fan space interface, used in Sonic.

**Comment by vadimp-nvidia** [comment]
In upstream line cards are powering and enabling happen automatically by "mlxreg-lc" driver. In Sonic OS these actions are commented out , because line cards powering and enabling in Sonic are performed in user space according to OS internal decision after relevant events are received through udev netlink. This is why access to hardware is commented out.

**Comment by alexrallen** [comment]
@saiarcot895 I'll try to fix it but its not trivial unfortunately with the tooling I have. 

@vadimp-nvidia would you be able to provide me with a list of associations for the kconfig for what sensors are intended for what platforms?

**Comment by vadimp-nvidia** [comment]
Which kconfig options have been added in this pull request, excepting MLXREG_LC? @alexrallen, could you, please point me out?
Each new config option, like MLXREG_LC, is coming with commit introducing new module, which is configured by this option:
https://github.com/torvalds/linux/commit/62f9529b8d5c87b89c3d9698405014bc8f370fbd#diff-25eaae9d15cf1d1d828afba8843eb0bc19cc83f36cd60c7b645326250e9ae941R37

**Comment by paulmenzel** [comment]
[…]

> This PR as a whole will be squash-merged as a single commit, so I'm fine with the individual commits here fixing things up.

Please do not do that. Maybe the patches could be done in one commit, but the Kconfig changes definitely should be in a separate commit with understandable commit messages.

**Comment by vadimp-nvidia** [comment]
OK, I didn’t realize that we are talking about changes in config file itself, and not about new kernel config options, added by new code.
I see it in:
https://github.com/Azure/sonic-linux-kernel/pull/259/commits/87425b60870630808d1790e0de86760307e65d14
Of course, new added options to kconfig should be in separate patch.
Below is the purpose of added options.

The below pin control options are necessary for new Nvidia systems based on CoffeLake COME to allow GPIO configuration:
CONFIG_PINCTRL_INTEL=m
CONFIG_PINCTRL_CANNONLAKE=m
CONFIG_PINCTRL_DENVERTON=m

Theses options are used be new coming SN2201 system (48x1G, 4x1000G):
CONFIG_I2C_MUX_PCA954x=m (I2C switch device, used to create mux infra)
CONFIG_GPIO_PCA953X=m (I2C GPIO expander, provided access for LED control)

PCIe switch ambient temperature sensor, used by SN4800 modular system
CONFIG_SENSORS_TMP421=m:

Ambient system sensor, used on several Spectrum-3 based systems as second source for TMP102 device: 
CONFIG_SENSORS_STTS751=m

Voltage and power monitoring devices used for 
CONFIG_SENSORS_ADS1015=m
CONFIG_SENSORS_POWR1220=m
CONFIG_TI_ADS1015=m

Fan driver for SN2201 system
CONFIG_SENSORS_EMC2305=m

**Comment by dgsudharsan** [comment]
> 

In SONiC all the commits are squashed and merged from a PR. I am not sure why is there is a specific request on this PR. If this is an overall policy change that is proposed please suggest it in the sonic community and make it as a standard methodology. We make our changes based on the fact that at the end there will be a single commit and this is the current process followed.

**Comment by dgsudharsan** [comment]
> 

In SONiC all the commits are squashed and merged from a PR. I am not sure why is there is a specific request on this PR. If this is an overall policy change that is proposed please suggest it in the sonic community and make it as a standard methodology. We make our changes based on the fact that at the end there will be a single commit and this is the current process followed.

**Comment by alexrallen** [comment]
@saiarcot895 @paulmenzel 

I have refactored this PR into three commits.

1. Reformatting existing patches
2. Adding newly backported patches
3. Updating the kconfig (with justification in the commit message)

I was also able to clean up the naming of the patches a bit, I hope this is now acceptable.

**Comment by saiarcot895** [comment]
That works for me.

If the commits are organized/logically separated, then I may do a regular merge; otherwise, the default for SONiC is to do a squash merge. In this case, a regular merge will be done here.

**Pull request merged**